### DISCUSSION TITLE
Detect panel fail to change state

### DISCUSSION THREAD

**Discussion by AdarWa** [davesmeghead/visonic#174] [General]
Hey,
First of all, I want to clarify that I don't use this integration as I am using a power link and I am doing only basic stuff, so I am making my own integration which is heavily based on this one.

I am looking to detect when the panel fail to change it's state(when you try to arm the system but something is open) from the KP-241 Remote Control.

Adding the PDUs and data chunks for when I try to arm but fail.
```
['0d', 'b0', '03', '24', '1a', 'ff', '08', 'ff', '15', '02', '00', '00', '00', '00', '00', '00', '00', '09', '08', '01', '03', '08', '00', '14', '05', '01', '00', '86', '00', '00', '26', '43', 'c8', '0a']
[datasize 8  index 255   length 21    data [2, 0, 0, 0, 0, 0, 0, 0, 9, 8, 1, 3, 8, 0, 20, 5, 1, 0, 134, 0, 0]]

['0d', 'a5', '00', '04', '00', '46', '00', '00', '00', '02', '00', '00', '43', 'cb', '0a']

['0d', 'a5', '00', '04', '00', '46', '00', '00', '00', '02', '00', '00', '43', 'cb', '0a']

['0d', 'b0', '03', '39', '06', 'ff', '08', 'ff', '01', '24', '27', '43', '76', '0a']
[datasize 8  index 255   length 1    data [36]]

['0d', 'b0', '03', '51', '06', 'ff', '08', 'ff', '01', '24', '28', '43', '5c', '0a']
[datasize 8  index 255   length 1    data [36]]

['0d', 'b0', '03', '24', '1a', 'ff', '08', 'ff', '15', '02', '00', '00', '00', '00', '00', '00', '00', '0a']
[datasize 8  index 255   length 21    data [2, 0, 0, 0, 0, 0, 0]]

['0d', 'b0', '03', '54', '29', 'ff', '48', 'ff', '24', '03', '2a', '08', '00', '01', '14', '00', '07', '02', '03', '2a', '15', '00', '01', '14', '00', '14', '02', '04', '02', '01', '00', '01', '2c', '00', 'ff', '02', '04', '02', '01', '00', '01', '2b', '00', 'ff', '02', '2a', '43', 'c8', '0a']
[datasize 72  index 255   length 36    data [3, 42, 8, 0, 1, 20, 0, 7, 2, 3, 42, 21, 0, 1, 20, 0, 20, 2, 4, 2, 1, 0, 1, 44, 0, 255, 2, 4, 2, 1, 0, 1, 43, 0, 255, 2]]
```

Would really appreciate if someone would help me with that.
@davesmeghead I hope you don't mind that I'm mentioning you as I know that you know this alarm system from the inside out :smile:

**Comment by davesmeghead** [comment]
Hi, I assume that you have a PowerMaster panel as you're listing `B0` message data.  Powerlink 3.1 hardware is supported by this integration using Mark's proxy server described here and on his github page
https://community.home-assistant.io/t/visonic-powermax-and-powermaster-integration/316702/572
https://github.com/[msp1974/visonic_proxy](https://github.com/msp1974/visonic_proxy)

The proxy supports both a simple bridge between the panel and my integration or a 3 way that also includes an external Visonic Server, when set up correctly.

On to your specific question. You know in advance that the panel will not arm as the panel is not "Ready", this is obtained through the `'0d', 'b0', '03', '24'` message.  I suggest that you look at my dev release or Mark's code for more details of decoding this message.
I don't specifically know of a message that contains the reason for not arming, that's why I create an Event that lists the open sensors, if any.

EDIT: The last sentence is not quite true, for things like invalid pin codes the panel sends an "Access Denied" message

**Comment by AdarWa** [comment]
Hey!
Thank you for the useful response!
I need to trigger an HA automation when trying to arm the panel, but fail to so checking the readiness won't work as it sends this command at other occasions too.
I've found a PDU which the alarm sends which I have no idea it it does.
`'0d', 'b0', '03', '54' ` I haven't found anything in the your code that regards this PDU😀

Thank you again 🙏

**Comment by AdarWa** [comment]
Hey!
Thank you for the useful response!
I need to trigger an HA automation when trying to arm the panel, but fail to so checking the readiness won't work as it sends this command at other occasions too.
I've found a PDU which the alarm sends which I have no idea it it does.
`'0d', 'b0', '03', '54' ` I haven't found anything in the your code that regards this PDU😀

Thank you again 🙏

**Comment by davesmeghead** [comment]
Hi,

As far as I've seen, if you try to arm the panel using the integration, the panel's keypad or any fobs etc, and the panel fails to arm with the voice message "not able to arm" etc, then it does not send a message to my integration.  As least, I have not seen any message that is coincident with this activity that I can reliably confirm.

Having said that, there are many `B0` message types that I cannot yet decode even partially, and it may be that a `54` data packet has the needed information.

If you manage to reliably work out any message content then please share it with the community here and I'll incorporate it in to the codebase.
